### DISCUSSION TITLE
Replace xref macros with Markdown links

### DISCUSSION THREAD

**Discussion by wbamberg** [mdn/mdn-community#624] [Content]
Over in https://github.com/mdn/content/pull/13802 , @queengooborg replaced all usages of the [`{{anch}}`](https://github.com/mdn/yari/blob/main/kumascript/macros/anch.ejs) macro with Markdown links.

I'd like to propose that we do the same with our xref macros - that is, replace them with plain Markdown links.

## What are xref macros?

These are a class of KumaScript macro that create a link, usually within MDN, like `{{domxref}}` or `{{cssxref}}` or `{{jsxref}}`, but sometimes outside MDN, like `{{interwiki}}` or `{{rfc}}`. When we did a census of KumaScript macros a few years ago (https://docs.google.com/document/d/1G5KwoJNi8mLmbzcWo-eawH0xw4spVvHGY4Wc4ZEuMYg/edit) we counted 50 of them, making them the second-largest category of macro. They are probably the largest now, since the old largest category of "Internal" has been reduced in the meantime.

## Arguments against keeping them

All other things being equal it's always better to use normal Markdown for things rather than macros.

* Contributors can use Markdown links without having to learn special stuff about MDN
* Removing macros simplifies the MDN platform
* Using more plain Markdown makes MDN content more portable
* Our xref macros have inconsistent APIs and even MDN experts have to look up the details. For instance, `{{domxref}}` takes a fourth argument that will suppress code formatting, while `{{cssxref}}` AFAIK does not let you suppress code formatting.

## Arguments in favour of keeping them

* They automatically add code formatting. This is a genuine benefit most of the time, although we don't always want code formatting: we often see things like `{{domxref("Fetch_API")}}` which incorrectly code-formats the text. Here the writer needs to do `{{domxref("Fetch_API", "Fetch API", "", "something")}}` which isn't at all obvious.

* They're quicker to write and easier to read. This is sometimes true. `{{domxref("Request")}}` looks great. But in many cases it breaks down, especially with docsets that don't have such a simple organization. For example, `{{jsxref}}` works pretty well for global objects, but:
    * to create a link to [`for...of`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of) you have to do `{{jsxref("Statements/for...of", "for...of")}}`
    * to create a link to [`https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await`]() you have to do `{{jsxref("Operators/await", "await")}}`
    * to create a link to [`Intl.NumberFormat.format()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat/format) you have to do `{{jsxref("Intl/NumberFormat/format", "Intl.NumberFormat.format()")}}`. This is very non-obvious!

**Comment by meyerweb** [comment]
I’ll add an argument in favor of keeping them: not having to worry about the URL structure of the links, which is nice since the URL structure is often divorced from the directory structure.  I used these a lot when doing the Temporal documentation, and having the system auto-link things was very nice.  It made find-and-replace of terms with `{{jsxref("\1")}}` a breeze.

**Reply by wbamberg** [reply]
One of the things though is that you do (often) still have to worry about URL structure. That's why you have to write things like `{{jsxref("Statements/for...of", "for...of")}}` - because that page is under "Statements", not "Global_objects".

**Comment by teoli2003** [comment]
I think there are different cases here: the xref macros that link to outside pages are doing very few things: `{{interwiki}}` for example is more complicated to use (and rare) so that writers add Markdown links today rather than using it.

`{{rfc}}` has for only advantage to link to the "right" site: RFCs are available in different formats and this uniformize it. But this goal is not achieved as a lot of RFC links are direct, so I think we can remove it (and update the links). (Can we add a Github test checking for sites that we don't want to link to here? It won't be exhaustive but can likely solve 99% of the cases)

For `{{domxref}}`, `{{cssxref}}`, …, they are very convenient for writers as they act as shorthands that are quicker to type and less error-prone. Maybe the solution here is to remove them but to provide an MDN companion add-on for VSCode or other editors to simplify the creation of these links.

**Reply by wbamberg** [reply]
I guess I just don't agree that they are easier.

For instance, I'm just looking at a PR that wants to link to https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/reset . They have tried using `{{domxref}}`, fairly naturally, and of course that doesn't work. So I try to remember the name of the HTML xref, and look it up - turns out it's called `HTMLElement`  not `htmlxref` or something consistent. Then I have to think, well, https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/reset is not directly under "Element". So will it work if I do `{{HTMLElement("input/reset", "<input type=\"reset\">")}}`?

This seems to interpret the argument as HTML, so I get:

********

<img width="310" alt="Screen Shot 2022-03-18 at 9 07 30 PM" src="https://user-images.githubusercontent.com/432915/159106048-ec6ed1d6-ae7a-4bfe-9350-db52f034fa3a.png">

********

It turns out that I can get the right effect with: `{{HTMLElement("input/reset", "<code>&lt;input type=\"reset\"></code>")}}`.

Or instead of all that, I could have written:

    [`<input type="reset">`](/en-US/docs/Web/HTML/Element/input/reset)

...and got a link which I know will work, and which will always work in exactly the same way, totally predictably, wherever I want to link to.

I'm pretty experienced with MDN and find myself doing things like this all the time with xrefs.

**Reply by hamishwillee** [reply]
IMO we would be better off moving to a normal link format.
I like the macros, but overall I think that "most users" would be better off if we only had one way to do links.

What I don't like about the link format is that it is less succinct, doesn't link in github, and must be translated because `/en-US` is part of the text. Would be good to look at our URL format and see if we could address that before making any kind of global change. 

No idea how, but some ideas:
- Allow language to be inferred - i.e. default URL format could be.
   ```
   [`<input type="reset">`](docs/Web/HTML/Element/input/reset)
   ```
- Accept relative links, so under if in tree under /Element you could specify
  ```
  [`<input type="reset">`](input/reset)
  ```
- Allow full URL to be specified and yari turns it into the right thing
  ```
  [`<input type="reset">`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/reset)
  ```
  I like this because the instructions are "just copy the URL", but it isn't as succinct.
- Do "whatever" to make the links work on github - that allows the links to be tested from github.
- Add tools (VS code?) for API link autocomplete. Further, these might use link reference syntax
- I'd love to suggest using markdown link reference format like below so we could perhaps have a succint format for the links "in-text"
  ```
  // Link to a reference like this
  [`logo`]
  
  // Declare the target like this
  [`logo`]: /en-US/docs/Web/HTML/Element/input/reset
  ```
  But it won't work, since you can't style the link text and still link.

**Reply by MrBrain295** [reply]
@hamishwillee linking within GitHub works but it has to be exactly the same as the actual directory tree (so no `/docs` and the case needs to be the same).

**Reply by hamishwillee** [reply]
@MrBrain295 Indeed. My point is right now the format includes /docs etc, so this does not work in github.

**Reply by teoli2003** [reply]
Note that @hamishwillee would be useful, independently of the removal of the xref macros, and would likely help.

I would avoid allowing `../` though, I think it makes links difficult to read. (I know this is not proposed, but just in case)

**Reply by hamishwillee** [reply]
@teoli2003 FWIW the big problem with relative links like `../` is that they only work well if you have a very rigid file system structure and you can always predict the indirection. Otherwise you have to think about what indirection you need for every single link, which is something that is hard to review and ensure you got right. So yes, let's avoid that if we can and make it easier for authors.

**Reply by Rumyra** [reply]
I've gone back and forth with how I feel about keeping or removing xref's - I wanted them because it _is_ easier as an author a lot of the time, however, this is true because I know them... now I consider the 'Arguments against keeping them' - primarily contributor ease of authoring. 

> I'd love to suggest using markdown link reference format like below so we could perhaps have a succinct format for the links "in-text"

I think a markdown solution could be an interesting approach, but as Hamish says this can prove tricky when it comes to styling.
